### DISCUSSION TITLE
Previous value of an Atom

### DISCUSSION THREAD

**Discussion by slimshader** [codewriter-packages/UniMob#6] [Q&A] [answered]
Hey, Is there a way to get previous value of an [Atom] property inside of Reaction?

**Comment by vanifatovvlad** [comment, marked as answer]
No, there is no way get the previous value of an atom. Alternatively, you can write reaction that remembers the previous value. 

```csharp
private void Start()
{
    Reaction(
        () => Counter,
        (current, previous) => Debug.Log($"Counter changed from {previous} to {current}")
    );
}

private static Reaction Reaction<T>(AtomPull<T> reaction, Action<T, T> effect)
{
    var hasPrevious = false;
    var previous = default(T);
    return Atom.Reaction(reaction, value =>
    {
        if (hasPrevious)
        {
            effect(value, previous);
        }

        previous = value;
        hasPrevious = true;
    });
}
```

**Comment by slimshader** [comment]
> No, there is no way get the previous value of an atom. Alternatively, you can write reaction that remembers the previous value.
> 
> ```cs
> private void Start()
> {
>     Reaction(
>         () => Counter,
>         (current, previous) => Debug.Log($"Counter changed from {previous} to {current}")
>     );
> }
> 
> private static Reaction Reaction<T>(AtomPull<T> reaction, Action<T, T> effect)
> {
>     var hasPrevious = false;
>     var previous = default(T);
>     return Atom.Reaction(reaction, value =>
>     {
>         if (hasPrevious)
>         {
>             effect(value, previous);
>         }
> 
>         previous = value;
>         hasPrevious = true;
>     });
> }
> ```

Thanks! Btw, I want to use it to find the difference of 2 collections so that the view layer could efficiently update it's contents. I am trying to replace UniRx-based view models with [Atom]-based.
